### PR TITLE
Script command for enchantgrade

### PR DESCRIPTION
**Pull request by Atemo**
* **Addressed Issue(s)**: N/A

* **Server Mode**: re

* **Description of Pull Request**: 

Added several commands to manage the enchantment grade of equipment.
- getitem4
- getitembound4
- rentitem4
- makeitem4
- delitem4
- countitem4
- rentalcountitem4

### CODE REVIEW

File: src/map/atcommand.cpp [hunk 1]
```diff
@@ -2429,6 +2429,82 @@ ACMD_FUNC(refine)
...
+		sprintf(atcmd_output, msg_txt(sd,1004), EQP_ACC_R); // %d: Right Accessory
+		clif_displaymessage(fd, atcmd_output);
+		sprintf(atcmd_output, msg_txt(sd,1005), EQP_HEAD_TOP); // %d: Top Headgear
+		clif_displaymessage(fd, atcmd_output);
+		sprintf(atcmd_output, msg_txt(sd,1006), EQP_HEAD_MID); // %d: Mid Headgear
+		clif_displaymessage(fd, atcmd_output);
+		return -1;
+	}
+
+	grade = cap_value(grade, -ENCHANTGRADE_A, ENCHANTGRADE_A);
```
Comment (Lemongrass3110) [inline src/map/atcommand.cpp:2468]: MAX_ENCHANTGRADE

File: src/map/atcommand.cpp [hunk 2]
```diff
@@ -2429,6 +2429,82 @@ ACMD_FUNC(refine)
...
+			continue;
+		if(j == EQI_AMMO)
+			continue;
+		if (pc_is_same_equip_index((enum equip_index)j, sd->equip_index, i))
+			continue;
+
+		if(position && !(sd->inventory.u.items_inventory[i].equip & position))
+			continue;
+
+		final_grade = cap_value(sd->inventory.u.items_inventory[i].enchantgrade + grade, 0, ENCHANTGRADE_A);
```
Comment (Lemongrass3110) [inline src/map/atcommand.cpp:2483]: MAX_ENCHANTGRADE

File: src/map/atcommand.cpp [hunk 3]
```diff
@@ -2429,6 +2429,82 @@ ACMD_FUNC(refine)
...
+	}
+
+	if (count == 0)
+		clif_displaymessage(fd, msg_txt(sd,1520)); // No item has been upgraded.
+	else if (count == 1)
+		clif_displaymessage(fd, msg_txt(sd,1521)); // 1 item has been upgraded.
+	else {
+		sprintf(atcmd_output, msg_txt(sd,1522), count); // %d items have been upgraded.
+		clif_displaymessage(fd, atcmd_output);
+	}
```
Comment (Lemongrass3110) [inline src/map/atcommand.cpp:2503]: I know this is probably a copy of `@refine` but is it really necessary to make a difference like that?
`// %d items have been upgraded.` would be good enough for all cases, at least for me.

File: conf/msg_conf/map_msg.conf [hunk 4]
```diff
@@ -1786,5 +1786,11 @@
 1517: Cannot add fame to class '%s'.
 1518: %d points were added to '%s'.
 
+// @grade
+1519: Please enter a position and an amount (usage: @grade <equip position> <+/- amount>).
+1520: No item has been upgraded.
+1521: 1 item has been upgraded.
+1522: %d items have been upgraded.
```
Comment (Lemongrass3110) [inline conf/msg_conf/map_msg.conf:1793]:
```suggestion
1520: No item has been upgraded.
1521: 1 item has been upgraded.
1522: %d items have been upgraded.
```
upgraded -> graded

File: src/map/script.cpp [hunk 5]
```diff
@@ -8533,9 +8623,24 @@ BUILDIN_FUNC(delitem2)
...
+	uint8 flag = 0x1;
+
+	if (command[strlen(command)-1] == '3' || command[strlen(command)-1] == '4') {
+		int offset = 11;
+
+		if (command[strlen(command)-1] == '4') {
+			offset = 12;
+
+			int grade = script_getnum(st, 11);
+			if (grade < ENCHANTGRADE_NONE || grade > ENCHANTGRADE_A) {
```
Comment (Lemongrass3110) [inline src/map/script.cpp:8635]: MAX_ENCHANTGRADE

File: src/map/script.cpp [hunk 6]
```diff
@@ -8116,14 +8189,29 @@ BUILDIN_FUNC(makeitem2) {
...
+		bool canShowEffect = false;
+
+		if (funcname[strlen(funcname)-1] == '3' || funcname[strlen(funcname)-1] == '4') {
+			int offset = 14;
+
+			if (funcname[strlen(funcname)-1] == '4') {
+				offset = 15;
+
+				int grade = script_getnum(st, 14);
+				if (grade < ENCHANTGRADE_NONE || grade > ENCHANTGRADE_A) {
```
Comment (Lemongrass3110) [inline src/map/script.cpp:8201]: MAX_ENCHANTGRADE

File: src/map/script.cpp [hunk 7]
```diff
@@ -7876,11 +7934,25 @@ BUILDIN_FUNC(rentitem2) {
...
+		if (!res)
+			return SCRIPT_CMD_FAILURE;
+	}
+	else if (funcname[strlen(funcname)-1] == '4') {
+		bool res = script_getitem_randomoption(st, sd, &it, funcname, 12);
+		if (!res)
+			return SCRIPT_CMD_FAILURE;
+
+		int grade = script_getnum(st, 11);
+		if (grade < ENCHANTGRADE_NONE || grade > ENCHANTGRADE_A) {
```
Comment (Lemongrass3110) [inline src/map/script.cpp:7947]: MAX_ENCHANTGRADE

File: src/map/script.cpp [hunk 8]
```diff
@@ -7715,19 +7759,30 @@ BUILDIN_FUNC(getitem2)
...
 		if (offset != 0) {
-			int res = script_getitem_randomoption(st, sd, &item_tmp, command, offset);
-			if (res == SCRIPT_CMD_FAILURE)
+			bool res = script_getitem_randomoption(st, sd, &item_tmp, command, offset);
+			if (!res)
 				return SCRIPT_CMD_FAILURE;
+
+			if (grade_offset > 0) {
+				int grade_tmp = script_getnum(st, grade_offset);
+				if (grade_tmp < ENCHANTGRADE_NONE || grade_tmp > ENCHANTGRADE_A) {
```
Comment (Lemongrass3110) [inline src/map/script.cpp:7768]: MAX_ENCHANTGRADE

**Review by Lemongrass3110 — COMMENTED** [review]
Shit is getting out of hands with all the item commands.
Maybe we should think of a better long term solution... x_x

File: src/map/atcommand.cpp (hunk 3 above)
Comment (Atemo) [reply src/map/atcommand.cpp]: Suggestion approved !

**Comment by gravity-ro** [comment]
[Added grade atcommand](https://github.com/rathena/rathena/pull/7180/commits/1a80501a3f713f1d2f5667c18f198bc60a8b587f)

Map Server Error

Screenshot_1

Screenshot_1x

Screenshot_2

How i can fix Thank you.

**Pull request merged**